### ISSUE TITLE
Infinite loop in plex_tokenize

### ISSUE THREAD

**Issue by FreeBASIC-programmer** [phorward/phorward#2]
The function plex_tokenize contains an infinite loop. The infinite loop looks like this
```
while( start )
	{
		if( ( start = plex_next( lex, start, &id, &end ) ) && matches )
		{
			if( ! *matches )
				*matches = parray_create( sizeof( prange ), 0 );

			r = (prange*)parray_malloc( *matches );
			r->id = id;
			r->begin = start;
			r->end = end;
		}

		start = end;
		count++;
	}
```
At end-of-input plex_next will return 0x0 and start becomes 0x0.  But the following statement 
```
               start = end;
```
assigns  a value to start that's never 0x0.  Which means start never becomes 0 and the loop never finishes.

When there is no more input *end is 0 but end itself contains the address of a character (which happens to be \0). *end is 0 but end is not. 

I found the error while testing a scanner for the C programming language. Using the function plex_lex I was able to tokenize a file from the phorward library (pregex.c) without a problem. But when I tried to do the same using plex_tokenize I ran into the issue described above. 

It's great fun playing around with plex and the. Thanks for creating the Phorward Toolkit.

Aside: I found a typo in the latest unicc documentation (it's in the index):

> 2.6. Implementing a comiler..................................................................................................................26

comiler => compiler

I found some more typos in the unicc manual. I'll read the entire manual and then post all the typos I find at https://github.com/phorward/unicc

**Comment by phorward** [comment]
Hi @FreeBASIC-programmer, thank you very much for your nice words 👍
Yep I already ran into this issue when implementing the plex command-line tool. It uses a similar loop, and also got into an infitine loop. But I didn't fix it here, so thanks for your report. It now runs well, I think... pleae check out 569fd40086b978d975c904bf81972e86c296f4c8. 

Getting some help with unicc would also be nice. Soon I would like to start UniCC v2 as a merge of the developments done in libphorward so far and the newer things in UniCC. I think that's the right step. And ideas grew, I still want to develop a GLR parser from this... check out the draft notices in the UniCC wiki at https://github.com/phorward/unicc/wiki/UniCC-v2.

Please keep up your great support, also with the manual would be so nice, I really appreciate this!

So long,
Jan

**Comment by phorward** [comment]
Oh right, this is the test program that I used:

```c
#include "phorward.h"

void printitem( prange* r )
{
	printf( "%d >%.*s<\n", r->id, r->end - r->begin, r->begin );
}


int main()
{
	int		cnt;
	plex*	l;
	parray*	a;
	char*	s = "1 23 456 789   ";

	l = plex_create(0);
	plex_define( l, "\\d+", 1, 0 );
	plex_prepare( l );

	cnt = plex_tokenize( l, s, &a ); 
	printf("=%d\n", cnt );

	parray_iter( a, (parrayfn)printitem);
}
```

**Comment by FreeBASIC-programmer** [comment]
I have downloaded/compiled https://github.com/phorward/phorward/commit/569fd40086b978d975c904bf81972e86c296f4c8 and tested function plex_tokenize. And it worked as expected. 
Repeatedly calling plex_lex (once for every token) as opposed to calling plex_tokenize once does not have an impact on runtime performance. I managed to tokenize an entire C source code file using plex_tokenize.
Some stats:
input file : https://github.com/phorward/phorward/blob/develop/src/regex/regex.c
time to tokenize the input file: round about 0.07 seconds.
number of tokens found: 5265
number of patterns defined using plex_define: 102
(I defined patterns for comments, newlines, whitespace, keywords, strings etc... etc...)
